design/nodes-machines-and-hosts.md

```diff
@@ -0,0 +1,213 @@
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# nodes-machines-and-hosts
+
+## Status
+
+implementable
+
+## Table of Contents
+
+<!--ts-->
+   * [nodes-machines-and-hosts](#nodes-machines-and-hosts)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+            * [Data Model](#data-model)
+            * [Growing the Cluster](#growing-the-cluster)
+            * [Shrinking the Cluster](#shrinking-the-cluster)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives](#alternatives)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-20T17:58-05:00 -->
+
+<!--te-->
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This document explains the use of Node, Machine, and BareMetalHost
+objects to represent different facets of implementing a self-hosted
+and Kubernetes-native hardware provisioning system. It explains the
+role of each, and why we are using 3 separate resources.
+
+## Motivation
+
+This document details the original design decisions made when the
+MetalKube project was launched, and places the information in context
+as an aid to future contributors.
+
+### Goals
+
+- Explain Node, Machine, and BareMetalHost and how they relate to
+  MetalKube.
+- Encourage input on that design, to refine it.
+
+### Non-Goals
+
+- Discussion of user interface designs related to the resources.
+
+## Proposal
+
+The MetalKube project is building a hardware provisioning system with
+a Kubernetes-native API. It includes a provider implementation that
+fits into the API defined by the cluster-api SIG. It differs from
+other similar implementations by being self-hosted. Where the AWS or
+OpenStack providers run within Kubernetes, they allocate resources
+from a pool *outside* of the Kubernetes cluster. MetalKube will use
+Kubernetes for tracking available resources.
+
+### Implementation Details/Notes/Constraints
+
+#### Data Model
+
+`Node` objects represent a running instance of kubelet. Their status
+fields include basic information about the health of the environment
+in which kubelet is running. A user manipulates the Node to change the
+cluster, such as draining the node for maintenance. The definition of
+Node is owned by the Kubernetes core, and offers few opportunities to
+attach custom data.
```

Comment:
Let's link to definitive sources of definitions where we can find them, as well.  For Nodes, I see https://kubernetes.io/docs/concepts/architecture/nodes/

I'm not sure the first few sentences there really help, but it does look like a good resource for diving into a lot more detail on the API.

Reply:
Done.